cachelib/allocator/CacheAllocator-inl.h

```diff
@@ -935,11 +959,19 @@ CacheAllocator<CacheTrait>::acquire(Item* it) {
 
   SCOPE_FAIL { stats_.numRefcountOverflow.inc(); };
 
-  if (LIKELY(incRef(*it))) {
-    return WriteHandle{it, *this};
-  } else {
-    // item is being evicted
-    return WriteHandle{};
+  while (true) {
```

Comment:
We discussed (1) in a sync. Writing down summary. TLDR: not an issue.

In order for us to unmark an item as moving, we must have removed it from the AccessContainer. `acquire()` also happens under AccessContainer lock (this is called from within a callback within AccessContainer). This means in order for us to find an item and successfully increment refcount, this item must be in either of the two states: (A) item is valid and has not been moved yet (B) this is a new item that had been moved from an older item.

Reply:
(1) We rely on the access container (AC) lock. Let me describe the scenario:
**Thread 1 (BG Move thread):**
1. Mark moving
2. Allocate new item
3. Copy data
4. Update MM containers
5. Update Access container (under AC lock)
6. Unmark moving
7. Notify waiters and remove context under (under Move lock)

**Client thread (lookup):**
1. Lookup in Access Container (under AC lock)
2. Acquire called under AC lock
3. a. Acquire called before Thread 1 step 5. In that case this thread will acquire Move lock and add itself to the wait context.
3. b. Acquire called after Thread 1 step 5. In that case this thread will get the new Item (already moved).

The `tryGetHandleWithWaitContextForMovingItem` might return `false` if the move failed and the item should be evicted (markForEvictionWhenMoving is going to be called). That is why we need this `while` loop - on the next iteration `incRef` will return `RefcountWithFlags::incResult::incFailedEviction`.

(2) I think there is no issue with that because the acquire called under AC lock.